castoredc_api/importer/async_helpers.py

```diff
@@ -0,0 +1,235 @@
+    auto_send: bool,
+    study: "CastorStudy",
+    client: httpx.AsyncClient,
+):
+    """Creates a survey package instance asynchronously."""
+    url = study.client.study_url + "/surveypackageinstance"
+    body = {
+        "survey_package_id": survey_package_id,
+        "record_id": record_id,
+        "ccr_patient_id": None,
+        "email_address": email_address,
+        "package_invitation_subject": None,
+        "package_invitation": None,
+        "auto_send": auto_send,
+        "auto_lock_on_finish": False,
+    }
+    response = await client.post(url=url, json=body)
+    response.raise_for_status()
+    return response.json()
+
+
+async def async_update_report_data(data: list, study: "CastorStudy") -> list:
+    """Updates the Castor EDC database with given report datapoints."""
+    # Split list to handle error when len(tasks) > max_connections
+    chunks = [
+        data[x : x + study.client.max_connections]
+        for x in range(0, len(data), study.client.max_connections)
+    ]
+    responses = []
+    # Create a new client for each chunk because of problem when tasks > max_connections
+    for idx, chunk in enumerate(chunks):
+        async with httpx.AsyncClient(
+            headers=study.client.headers,
+            timeout=study.client.timeout,
+            limits=study.client.limits,
+        ) as client:
+            tasks = [async_upload_report_data(item, client, study) for item in chunk]
+
+            # Show progress bar while handling responses
+            temp_responses = [
+                await response
+                for response in tqdm(
+                    asyncio.as_completed(tasks),
+                    total=len(tasks),
+                    desc=f"Async Uploading {idx + 1}/{len(chunks)}",
+                    file=sys.stdout,
+                )
+            ]
+            responses = responses + temp_responses
+    return responses
+
+
+async def async_upload_report_data(item, client, study):
+    """Coroutine to upload a single row of report data to a new report and handle the response."""
+    # Copy because we don't want to overwrite dict
+    feedback = copy.deepcopy(item["row"])
+    try:
+        instance = await async_create_report_instance(
+            report_id=item["report_id"],
+            record_id=item["row"]["record_id"],
+            report_name_custom=item["report_name"],
+            study=study,
+            client=client,
+        )
+        body = create_report_body(instance, item["row"], study, item["upload_datetime"])
+        url = (
+            study.client.study_url
+            + f"/record/{item['row']['record_id']}/data-point-collection/"
+            f"report-instance/{instance['id']}"
+        )
+        json = {"data": body, "common": item["common"]}
+        feedback = await async_upload_data(client, feedback, json, study, url)
+    except httpx.HTTPStatusError as error:
+        feedback["error"] = error.response.json()
+    return feedback
```

Comment:
This makes error handling more difficult for consumers of your library:

1. In case of a transport-level exception (connection refused, for example) the exception will bubble up to the caller.
2. For cases where the body is json-decodable (errors generated by our API's application code), they get this `feedback` dict with an error key back.
3. For cases where the body is not json-decodable (errors generated by our nginx proxy, such as gateway timeouts or the HTML-only response you'll receive when the entire system is down for maintenance) the `json()` call will generate some kind of exception.

Reply:
Thanks for the note.
I wanted to let the async post requests keep running if one or more lines of data were rejected by the API.
Would it be sufficient to:
- Try to add the json() and add some error handling and different error message if the json() call fails?
- Catch transport level exceptions (Request errors as defined in https://www.python-httpx.org/exceptions/) and add some information on those in the feedback dict?

Reply:
Updated Error handling by taking into account requesterrors and json decode errors.